### DISCUSSION TITLE
Plugin ID automatic generation

### DISCUSSION THREAD

**Discussion by imqdee** [LedgerHQ/ledger-fresh-management#49] [Ideas]
## Context

I propose a simple and automatic solution to solve the following problems:
- Today the author of a plugin is not identified in any way in the plugin store (you have to check the author of the commit)
- Today there is no management, even minimal, of versioning, we do not know yet how a contributor can replace the plugin he has already pushed with a new version of this plugin
- In Ledger Fresh, without a system that regulates point 2, we won't be able to know if we have to init an install transaction or a multicall remove-install transaction

Today the type of Plugin is naively defined as follows (see #32) 

```typescript
type Plugin = {
  id: string, // classhash
  signature: string,
  name: string,
  description?: string,
}
```

I suggest to extends it for this:

```typescript
type Plugin = {
  id: string,
  classhash: string,
  signature: string,
  name: string,
  description?: string,
}
```

Note that the difference lies in the addition of the `classhash` key. Before the `id` represented the `classhash`, after my proposal this is no longer the case, the `id` is now a real identification key.

## How does it work?

The id will be an autogenerated key when a plugin is accepted to be added in the list of plugins stamped by Ledger. The id will have the following properties:
- Its value will respect the following template: `<author_github_handle>:<plugin_name>`. For example, if I propose myself tomorrow a plugin called StarkSignerMulti, the id of the plugin will be qd-qd:starksignermulti (the name of the plugin use the kebab-case convention and the special characters are formatted as they are in a URL, the GitHub handle remains as is)
- The id is unique in the plugin list. If a plugin is pushed into the list and it shares the same id as a plugin already in the list, the latter is replaced by the newcomer

This solution, easy to implement, allows solving points 1 and 2 mentioned in the introduction of my message. Now, the author is clearly identified and it is possible for him (and only him) to replace a plugin he has already submitted with a new version, by submitting to the review a plugin with the same name.

## Why is it useful?

Thanks to this, we now have everything we need to inform the user that he can "update" a plugin. Why? Because by checking the list of plugins he has installed on his account, and the list of plugins in the plugin store, if an installed plugin shares the same id as a plugin in the plugin store but not the same class hash, it means that it is a new version submitted by the author. In this case, when the user wants to install, we will make a multicall of remove-install (remove old class hash, install new class hash). In all other cases, we will just do a classic install transaction.

## Weakness points

The idea seems to me to be a very good compromise between dev time/security/versionning management, but it assumes that the Github handles remain static. So two concerns:
- The user can change his handle and thus not have the possibility to submit a new version of his plugin
- After a handle change, depending on the GitHub policy, another user could, perhaps, get the initial handle back to publish a new version of a plugin on behalf of

Point 1 almost never happens, it's a pain for the devs to change the handle. Point 2 is theoretical, I didn't dig into it, but since the plugins are reviewed by us anyway, the risk is probably small.

## Possible improvement to fix the weak points

In case we want to strengthen the system and counter these weak points, I propose that later (post-v1), we decide to change the id template to incorporate a signature. The id will follow this template: `<author_github_handle>:<plugin_name>:<signature>`. This signature which will have to be done manually by the plugin author will ensure that the "update" is done by the same author who pushed the previous version of the plugin (by comparing the signature public key in the registry with the signature public key associated with the new version, they will have to match to be validated). This problem will solve the two weaknesses listed above and allow the management of plugins by different Github accounts.
I think we can do without it for the release of Ledger Fresh.

**Comment by imqdee** [comment]
@rdubois-crypto pointed out that this idea opens the door to a downgrade attack.
Indeed, if the plugin store is compromised, it would be possible for an attacker to downgrade a plugin already installed on the wallet of a user (because the hash would be different, meaning the front end would inform the user that an upgrade is available).

It's up to us to see what we want to do with this information

**Comment by btchip** [comment]
Small suggestion - maybe change the template to <author_github_handle>:<plugin_name>:<version>. This would also help with a downgrade attack if the user is careful

**Comment by yogh333** [comment]
Hi,
I have few comments:

1. `Plugin` type defines a `signature` field: how is this signature computed ?
2. `id` field is defined as `<author_github_handle>:<plugin_name>`. Then, according to @qd-qd's proposal, a signature could be added in the `id` field in order to prevent malicious plugin version downgrade. I do not understand where this new (?) signature will be added in the so-called `id` template. Do you mean `<author_github_handle>:<plugin_name>:<signature>` ? In that case, what the difference with the signature already present in the `Plugin` type ?
